### ISSUE TITLE
`Style/MethodCallWithArgsParentheses` within the block of a macro

### ISSUE THREAD

**Issue by oneandonlymike** [rubocop/rubocop#10435]
When using the `Style/MethodCallWithArgsParentheses` cop, it does not apply to the block of a macro. Due to this, rails tests written in the `test "name" do` format instead of `def test_name` format will not enforce the parentheses around the assertion method calls. 

--------

## Expected behavior

Given this test:
```Ruby
# frozen_string_literal: true

require "test_helper"

class FooBarTest < ActiveSupport::TestCase
  test "block" do
    assert_equal "foo", "bar"
  end
end
```

I would expect Rubocop to enforce and autocorrect to add parentheses around the arguments of `assert_equals`. 

## Actual behavior

Rubocop does not return or autocorrect the parentheses. 

## Steps to reproduce the problem

Given this full example you will be able to see the difference of `assert_equal` not being autocorrected in one format of test versus the other:
```Ruby
# frozen_string_literal: true

require "test_helper"

class FooBarTest < ActiveSupport::TestCase
  def test_method
    assert_equal "foo", "bar"
  end

  test "block" do
    assert_equal "foo", "bar"
  end
end
```

## RuboCop version

```sh
1.25.0 (using Parser 3.1.0.0, rubocop-ast 1.15.1, running on ruby 3.1.0 arm64-darwin21)
  - rubocop-minitest 0.17.0
  - rubocop-performance 1.13.2
  - rubocop-rails 2.13.2
  - rubocop-sorbet 0.6.3
```

---

If there is reason to preserve this behaviour it would be nice to have a configurable option to not have it behave this way.

**Comment by nobuyo** [comment]
I don't know if this will solve the problem, I found an option `IncludedMacros`.

```yml
# rubocop.yml

Style/MethodCallWithArgsParentheses:
  Enabled: true
  IncludedMacros: ['assert_equal']
```

If specify macro names to the option, autocorrect will add parentheses as follows:

```ruby
# frozen_string_literal: true

require 'test_helper'

class FooBarTest < ActiveSupport::TestCase
  def test_method
    assert_equal('foo', 'bar')
  end

  test 'block' do
    assert_equal('foo', 'bar')
  end
end
```

That said, it might be easier if there is an option to force parentheses on every call.

**Comment by dvandersluis** [comment]
The cop already has `AllowedPatterns` so adding this to your config should solve your issue:

```
Style/MethodCallWithArgsParentheses:
  AllowedPatterns:
  - '^assert'
  - '^refute'
```

**Comment by nobuyo** [comment]
I think `AllowedPatterns` does not solve the problem. Since `AllowedPattern` is an option that allows patterns that do not conform to cop, so the behavior on `EnforcedStyle: require_parentheses` would be the opposite.

Running `rubocop -a` with the previous configuration:
```
Style/MethodCallWithArgsParentheses:
  Enabled: true
  AllowedPatterns:
  - '^assert'
  - '^refute'
```

```rb
# frozen_string_literal: true

require 'test_helper'

class FooBarTest < ActiveSupport::TestCase
  def test_method
    assert_equal 'foo', 'bar'
    another_method 'baz', 'quux'
  end

  test 'block' do
    assert_equal 'foo', 'bar'
  end
end
```

is corrected to:

```rb
# frozen_string_literal: true

require 'test_helper'

class FooBarTest < ActiveSupport::TestCase
  def test_method
    assert_equal 'foo', 'bar'     # should be parenthesized, but didn't
    another_method('baz', 'quux') # parenthesized
  end

  test 'block' do
    assert_equal 'foo', 'bar'
  end
end

```

A new option like `ForceEligiblePatterns` is needed for this case IMO.

**Comment by dvandersluis** [comment]
Strangely, this cop works for `omit_parentheses` within a block but not for `require_parentheses`. I'll take a look at it.

**Comment by dvandersluis** [comment]
So the issue here for `require_parentheses` is the `IgnoreMacros` configuration. The methods inside the block are considered macros and with `IgnoreMacros` the parens aren't being required. I'm not really sure yet how we can fix this without breaking cases that other people are expecting.

**Comment by oneandonlymike** [comment]
> The methods inside the block are considered macros and with IgnoreMacros

This feels weird though with the difference being whether or not they are inside of a block.

I would expect the same corrections to happen regardless of whether I used `def test_` or `test "" do`

**Comment by dvandersluis** [comment]
> I would expect the same corrections to happen regardless of whether I used def test_ or test "" do

I understand that but at the same time a receiver-less method call within a block is explicitly treated as a macro everywhere by RuboCop. This is the same pattern:

```ruby
Rails.application.routes.draw do
  get :foo
end
```

@bbatsov what do you think we should do here?

**Comment by bbatsov** [comment]
The configurable escape hatch asked for here shipped in 1.82 as `IncludedMacroPatterns` (added via #14569, which was effectively a duplicate of this issue) - e.g. `IncludedMacroPatterns: ['\Aassert']` makes the cop flag those calls inside `test ... do` blocks too. There's also `IgnoreMacros: false` as the blunt version. Closing as resolved.
